keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,959 @@
+
+### CRD Schema Change Validation
+
+Create a CLI tool that evaluates the differences between an old and new CRD in YAML to identify changes that may break users.
+
+All validations are exported in Go packages that can be imported and consumed by other projects.
+
+Some examples of how the CLI may be used:
+
+- Comparing CRD that exists on cluster to a local YAML file of the same CRD - `crdify kube://mycrd.example.org file://mycrd.yaml`
+- Comparing CRD across git revisions (could be used in CI systems) - `crdify git://main?path=mycrd.yaml git://HEAD?path=mycrd.yaml`
+- Comparing CRD from git revision to a local file - `crdify git://main?path=mycrd.yaml file://mycrd.yaml`
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+*As an Operator Author, I want to ensure my Go APIs that are used to build my CustomResourceDefinition(s) follow Kubernetes API conventions and best practices*
+
+**How**: Using a golangci-lint plugin (or CLI) in both local development and CI environments to check for adherence to Kubernetes API conventions
+
+#### Story 2
+
+*As an API reviewer, I want to automate the process of finding deviations from Kubernetes API conventions and best practices so that I can focus on reviewing API changes on a more fundamental level*
+
+**How**: Using a golangci-lint plugin (or CLI) in both local development and CI environments to check for adherence to Kubernetes API conventions
+
+#### Story 3
+
+*As an Operator Author, I want to ensure that changes to my CustomResourceDefinition(s) will not break users*
+
+**How**: Using a CLI in both local development and CI environments to check for breaking changes in CRD schemas
+
+#### Story 4
+
+*As a Cluster Administrator, I want to validate that updating a CustomResourceDefinition won't break my cluster*
+
+**How**: Using a CLI in both local development and CI environments to check for breaking changes in CRD schemas
+
+#### Story 5
+
+*As a Kubernetes Package Manager maintainer, I want to identify when making updates to CustomResourceDefinition(s) may result in a breaking change so that we can alert users before making those changes*
+
+**How**: Using a library for identifying breaking changes in CRD schemas
+
+#### Story 6
+
+*As a Cluster Administrator, I want Kubernetes to prevent me from making breaking changes to CRDs on my cluster*
+
+**How**: Using an admission controller/plugin that rejects CRD update operations if it would result in a breaking change
+
+**Note for reviewers**: Adding an admission controller/plugin to enact this behavior is not part of this enhancement, but it is a use case that could be enabled by the introduction of the projects being proposed.
```

Comment:
We may want to add this to the non-goals then, we may also want an "Alternative" to explain why we aren't going this right now

Reply:
I can add this to the alternatives, although I may just update the non-goals to explain _why_ this is a non-goal of this KEP.